src/attention.jl

```diff
@@ -0,0 +1,125 @@
+const AA3{T} = AbstractArray{T,3}
+const AA4{T} = AbstractArray{T,4}
+const AA{N,T} = AbstractArray{T,N}
+
+"""
+    dot_product_attention(query, key, value; [bias, droput_fn, mask, num_heads])
+
+Multihead dot product attention used in transformer architectures. 
+
+The input arrays must have the first two dimensions given by the number of features
+and the sequece length, then an arbitrary number of batch dimensions or none. 
+
+# Arguments
+
+- `query`: Query array of size `(qk_dim, q_len, batch_size...)`.
+- `key`: Key array of size `(qk_dim, kv_len, batch_size...)`.
+- `value`: Value array of size `(v_dim, kv_len, batch_size...)`.
+- `bias`: Either `nothing` or an input array broadcastable to size `(kv_len, q_len, num_heads, batch_size)`. 
+          Can also be set to `mask=:causal` to apply a causal mask. Default `nothing`.
+- `dropout_fn`: A dropout function to apply on the attention scores. Default `nothing`. 
+- `mask`: Either `nothing` or an input array broadcastable to size `(kv_len, q_len, num_heads, batch_size)`. 
+          Can also be set to `mask=:causal` to apply a causal mask. Default `nothing`.
+- `num_heads`: Number of heads to split the input arrays into. Default `1`.
+
+# Examples
+    
+```julia
+q, k, v = rand(10, 20, 2), rand(10, 30, 2), rand(20, 30, 2)
+y, α = dot_product_attention(q, k, v)
+```
+"""
+function dot_product_attention(q::AA{N}, k::AA{N}, v::AA{N}; num_heads=1, kws...) where N
+    batch_size = size(q)[3:end]
+    
+    batch_size == size(k)[3:end] == size(v)[3:end] || throw(ArgumentError("Batch dimensions have to be the same."))
+    size(q, 1) == size(k, 1) || throw(ArgumentError("First dimension in query and key has to be the same."))
+    size(k, 2) == size(v, 2)  || throw(ArgumentError("Second dimension in key and value has to be the same."))
+    
+    q, k, v = map(x -> reshape(x, size(x, 1), size(x, 2), :), (q, k, v))
```

Comment:
Is this worth doing, or can we just start with the case taking only 3D arrays? 

If we do want to take 4D, then probably that should be a different method which reshapes before & after. Then e.g. CuArray can overload just the 3D one.

Reply:
No, this line does a bunch of reshaping that isn't necessary in the common case of 3D input. That's what seems confusing and I think would be better handled by the public function having two methods.

Reply:
done